### PR TITLE
Support `_id` aggregation operators for `$group` stage

### LINKED ISSUE #2694: Support recursive aggregation operators
### What should be done?

After https://github.com/FerretDB/FerretDB/issues/2680

The goal of this task is to support recursive aggregation operators in all supported operators (and accumulators) .

#### Operators support other operators
For example it should be possible to call:
`{$project: {$type: {$type: "foo"}}}`

All of the recursive operators should be tested with compat tests.

It might be necessary to implement better error handling inside `NewOperator` that would return failing operator for the message.

#### Accumulator operator support other operators
- `$group` stage `$sum` accumulator supports operators
example: `totalSaleAmount: { $sum: { $multiply: [ "$price", "$quantity" ] } }` from [manual](https://www.mongodb.com/docs/manual/reference/operator/aggregation/group/#group-by-day-of-the-year)
https://github.com/FerretDB/FerretDB/blob/0992e5f8a410e59041a62a12f666eacf6f3bd27b/internal/handlers/common/aggregations/operators/accumulators/sum.go#L57
Test passes
https://github.com/FerretDB/FerretDB/blob/bddfa9ee15fec28665fdc145bc1d22c7d7f4bf61/integration/aggregate_documents_compat_test.go#L1055-L1064
### Where? 

- `operators` package
- `accumulators` package
See the issue link in the code base.

### Definition of Done

- operators updated;
- integration/compatibility tests added/updated;
- spot refactorings done;
- multiple PRs (separate PRs for different operators).

### PR DESCRIPTION
**Pull request by noisersup**
# Description

Closes #2694.

## Readiness checklist

- [ ] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

**Comment by rumyantseva** [comment]
@noisersup the linter fails!

File: internal/handlers/common/aggregations/stages/group.go [hunk 1]
```diff
@@ -165,73 +171,102 @@ func (g *group) Process(ctx context.Context, iter types.DocumentsIterator, close
...
-			panic(fmt.Sprintf("unhandled field path error %s", exprErr.Error()))
+		var group groupMap
+
+		for _, doc := range in {
+			val, err := op.Process(doc)
+			if err != nil {
+				return nil, err
+			}
+
+			group.addOrAppend(val, doc)
```
Comment (rumyantseva) [inline internal/handlers/common/aggregations/stages/group.go:193]: Do you understand why this line is marked uncovered? It looks quite essential.

File: internal/handlers/common/aggregations/stages/group.go [hunk 2]
```diff
@@ -165,73 +171,102 @@ func (g *group) Process(ctx context.Context, iter types.DocumentsIterator, close
...
+		}
 
-	expression, err := aggregations.NewExpression(groupKey)
-	if err != nil {
-		var exprErr *aggregations.ExpressionError
-		if !errors.As(err, &exprErr) {
-			return nil, lazyerrors.Error(err)
+		op, err := operators.NewOperator(groupKey)
+		if err != nil {
+			return nil, err
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/stages/group.go:182]: Could you add a test for error case please? See how it's done for `$project` https://github.com/FerretDB/FerretDB/blob/4d534d602e82d7726f327c2915d73cc253195553/integration/aggregate_documents_test.go#L393-L428

File: internal/handlers/common/aggregations/stages/group.go [hunk 3]
```diff
@@ -165,73 +171,102 @@ func (g *group) Process(ctx context.Context, iter types.DocumentsIterator, close
...
-		// non-string key aggregates values of all `in` documents into one aggregated document.
-		return []groupedDocuments{{
-			groupID:   g.groupExpression,
-			documents: in,
-		}}, nil
-	}
+	switch groupKey := g.groupExpression.(type) {
+	case *types.Document:
+		if !operators.IsOperator(groupKey) {
+			break
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/stages/group.go:177]: I'm not sure why this is not covered by test :thinking: we should cover it.

File: internal/handlers/common/aggregations/stages/group.go [hunk 4]
```diff
@@ -165,73 +171,102 @@ func (g *group) Process(ctx context.Context, iter types.DocumentsIterator, close
...
-				"$group (stage)",
-			)
-		default:
-			panic(fmt.Sprintf("unhandled field path error %s", exprErr.Error()))
+		var group groupMap
+
+		for _, doc := range in {
+			val, err := op.Process(doc)
+			if err != nil {
+				return nil, err
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/stages/group.go:190]: Also we want to cover error case by test. When I was testing this with $sum with projection, I used recursive $sum to error like below
https://github.com/FerretDB/FerretDB/blob/4d534d602e82d7726f327c2915d73cc253195553/integration/aggregate_documents_compat_test.go#L1924

File: internal/handlers/common/aggregations/stages/group.go [hunk 5]
```diff
@@ -165,73 +171,102 @@ func (g *group) Process(ctx context.Context, iter types.DocumentsIterator, close
...
+			group.addOrAppend(val, doc)
 		}
-	}
 
-	var group groupMap
+		return group.docs, nil
 
-	for _, doc := range in {
-		val, err := expression.Evaluate(doc)
+	case *types.Array, float64, types.Binary, types.ObjectID, bool, time.Time, types.NullType,
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/stages/group.go:198]: I think we can handle them in default case :hugs:

**Review by chilagrow — COMMENTED** [review]
See comments about test coverage. Also there are failing tests on CI.

File: internal/handlers/common/aggregations/stages/group.go (hunk 5 above)
Comment (noisersup) [reply internal/handlers/common/aggregations/stages/group.go:199]: If I recall correctly, we prefer to list all of the types and panic if we approach an unexpected type, did that change?

File: internal/handlers/common/aggregations/stages/group.go [hunk 6]
```diff
@@ -265,6 +297,97 @@ func (m *groupMap) addOrAppend(groupKey any, docs ...*types.Document) {
...
+// - ErrFieldPathInvalidName when FieldPath is invalid.
+// - ErrNotImplemented when the operator or expression is not implemented yet.
+// - ErrOperatorWrongLenOfArgs when the operator has an invalid number of arguments.
+// - ErrInvalidPipelineOperator when the operator does not exist.
+// - ErrFailedToParse when operator has invalid variable expression.
+// - ErrGroupInvalidFieldPath when operator has empty path expression.
+func processOperatorError(err error) error {
+	if err == nil {
+		return nil
+	}
```
Comment (chilagrow) [inline internal/handlers/common/aggregations/stages/group.go:314]: err is always nil, so this check is not needed.

**Review by chilagrow — COMMENTED** [review]
Could you check the test coverage again please :pray: . There are still quite a bit that are not covered.

The error handing code seems quite duplicated, do you think we can merge them as stage input error handling?

File: internal/handlers/common/aggregations/stages/group.go [hunk 7]
```diff
@@ -265,6 +297,97 @@ func (m *groupMap) addOrAppend(groupKey any, docs ...*types.Document) {
 	})
 }
 
+// processOperatorError takes internal error related to operator evaluation and
+// returns proper CommandError that can be returned by $project aggregation stage.
+//
+// Command error codes:
+// - ErrEmptySubProject when operator value is empty.
```
[the remark this answers is not in the document]
Comment (noisersup) [reply internal/handlers/common/aggregations/stages/group.go:304]: This PR is still in progress, I didn't implement all of the changes yet but for sure everything about code coverage and comments will be changed :)

File: internal/handlers/common/aggregations/stages/group.go [hunk 8]
```diff
@@ -265,6 +297,97 @@ func (m *groupMap) addOrAppend(groupKey any, docs ...*types.Document) {
...
+// - ErrFieldPathInvalidName when FieldPath is invalid.
+// - ErrNotImplemented when the operator or expression is not implemented yet.
+// - ErrOperatorWrongLenOfArgs when the operator has an invalid number of arguments.
+// - ErrInvalidPipelineOperator when the operator does not exist.
+// - ErrFailedToParse when operator has invalid variable expression.
+// - ErrGroupInvalidFieldPath when operator has empty path expression.
+func processOperatorError(err error) error {
```
[the remark this answers is not in the document]
Comment (noisersup) [reply internal/handlers/common/aggregations/stages/group.go:321]: There are some other differences, I'd suggest to think about that after we implement proper operator handling for more stages, but I like the idea! :)

**Review by rumyantseva — APPROVED** [review]
LGTM

File: internal/handlers/common/aggregations/stages/group.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/common/aggregations/stages/group.go:305]: As we found out that error messages are slightly different so having standardized error handling is perhaps out of question :see_no_evil:

File: internal/handlers/common/aggregations/stages/group.go [hunk 9]
```diff
@@ -265,6 +297,97 @@ func (m *groupMap) addOrAppend(groupKey any, docs ...*types.Document) {
...
+				commonerrors.ErrNotImplemented,
+				"Invalid $group :: caused by :: "+opErr.Error(),
+				"$group (stage)",
+			)
+		case operators.ErrArgsInvalidLen:
+			return commonerrors.NewCommandErrorMsgWithArgument(
+				commonerrors.ErrOperatorWrongLenOfArgs,
+				"Invalid $group :: caused by :: "+opErr.Error(),
+				"$group (stage)",
+			)
```
[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/common/aggregations/stages/group.go:329]: Great!

**Review by chilagrow — APPROVED** [review]
Amazing work! :rocket:

**Pull request merged**